### PR TITLE
Fix DBCS attribute corruption during reflow

### LINKED ISSUE #12837: out of range exception reading UnicodeStorage during UpdatePatternLocations
1.12.10908

Uncertain as to repro steps.

Watson failure hash 326e3a1b-889b-f24e-582d-0e18d3c853b4

<html>
<body>

Frame | Name | | offset | | | | file || line
-- | -- | -- | -- | -- | -- | -- | -- | -- | --
0 | KERNELBASE!RaiseException |   | 0x000000000000006C |   |   |   | xcpt.c |   | 936
1 | VCRUNTIME140!_CxxThrowException |   | 0x0000000000000090 |   |   |   | throw.cpp |   | 75
2 | msvcp140!std::_Xout_of_range |   | 0x0000000000000022 |   |   |   | xthrow.cpp |   | 25
3 | Microsoft_Terminal_Control!CharRow::GetText |   | 0x000000000004414C |   |   |   | CharRow.cpp |   | 217
4 | Microsoft_Terminal_Control!TextBuffer::GetPatterns |   | 0x00000000000000C0 |   |   |   | textBuffer.cpp |   | 2665
5 | Microsoft_Terminal_Control!Microsoft::Terminal::Core::Terminal::UpdatePatternsUnderLock |   | 0x000000000000008E |   |   |   | Terminal.cpp |   | 1255
6 | Microsoft_Terminal_Control!winrt::Microsoft::Terminal::Control::implementation::ControlCore::UpdatePatternLocations |   | 0x0000000000000044 |   |   |   | ControlCore.cpp |   | 494

</body>
</html>

### PR DESCRIPTION
**Pull request by lhecker**
855e136 contains a regression which breaks buffer reflow if wide surrogate
characters are present. This happens because we made use of the
`TextBufferCellIterator` whose increment operator skips 2 cells for wide
characters. This created a "misalignment" in the reflow logic which was written
for cell-wise iteration. This commit fixes the issue, by reverting back to the
previous algorithm without iterators.

Closes #12837
Closes MSFT-38904421

## Validation Steps Performed
* Run ``pwsh -noprofile -command echo "`u{D83D}`u{DE43}"``
* Resizing conhost preserves all contents ✅
* Resizing Windows Terminal doesn't crash it ✅
* Added a test covering this issue ✅

### CODE REVIEW

**Review by miniksa — APPROVED** [review]

File: src/buffer/out/textBuffer.cpp [hunk 1]
```diff
@@ -2360,16 +2355,13 @@ HRESULT TextBuffer::Reflow(TextBuffer& oldBuffer,
...
-                const auto textAttr = *attrs;
+                const auto textAttr = row.GetAttrRow().GetAttrByColumn(copyAttrCol);
                 if (!newRow.GetAttrRow().SetAttrToEnd(newAttrColumn, textAttr))
                 {
                     break;
                 }
             }
             CATCH_LOG(); // Not worth dying over.
-
-            ++newAttrColumn;
```
Comment (DHowett) [inline src/buffer/out/textBuffer.cpp:2371]: It seems bad to not increment newAttrColumn now. We still use `SetAttrToEnd` on it`.

**Review by DHowett — CHANGES_REQUESTED** [review]
The newAttrColumn thing seems like it would regress the original bug we fixed.

**Comment by DHowett** [comment]
We will not erroneously copy a DBCS Trailing Attribute to the edge of the screen, right? Or a leading one?

**Comment by lhecker** [comment]
> We will not erroneously copy a DBCS Trailing Attribute to the edge of the screen, right? Or a leading one?

I'll have to check! This PR could've been a draft tbh. 😅
I'll need a bit to write a unit test and test this more thoroughly.

File: src/buffer/out/textBuffer.cpp [hunk 2]
```diff
@@ -2355,21 +2350,18 @@ HRESULT TextBuffer::Reflow(TextBuffer& oldBuffer,
...
         for (short copyAttrCol = iOldCol;
              copyAttrCol < cOldColsTotal && newAttrColumn < newWidth;
-             copyAttrCol++)
+             copyAttrCol++, newAttrColumn++)
         {
             try
             {
                 // TODO: MSFT: 19446208 - this should just use an iterator and the inserter...
-                const auto textAttr = *attrs;
+                const auto textAttr = row.GetAttrRow().GetAttrByColumn(copyAttrCol);
```
Comment (zadjii-msft) [inline src/buffer/out/textBuffer.cpp:2358]: sad that this beautiful optimization had to die

File: src/buffer/out/textBuffer.cpp [hunk 3]
```diff
@@ -2311,9 +2309,9 @@ HRESULT TextBuffer::Reflow(TextBuffer& oldBuffer,
             try
             {
                 // TODO: MSFT: 19446208 - this should just use an iterator and the inserter...
-                const auto glyph = chars->Char();
-                const auto dbcsAttr = chars->DbcsAttr();
-                const auto textAttr = *attrs;
+                const auto glyph = row.GetCharRow().GlyphAt(iOldCol);
+                const auto dbcsAttr = row.GetCharRow().DbcsAttrAt(iOldCol);
```
Comment (zadjii-msft) [inline src/buffer/out/textBuffer.cpp:2313]: ah cause `chars++` might skip two _columns_ for something like an emoji, right?

Comment (lhecker) [reply src/buffer/out/textBuffer.cpp:2313]: Yeah I think so.

**Comment by DHowett** [comment]
So, here is my hope.

WinDev is in "ask" mode locally, so I'm just gonna take this fix up during Ask for both Terminal and Console.
We need this to be *rock solid*. I need tests for resizing single- and double- width surrogate pairs, at the edge of the buffer, wider and narrower.
I need to understand the impact on Terminal, and I think we should wrap UpdatePatternLocationsUnderLock in a try/catch _just to be safe_ so that any future regressions here only break URL updating and don't cause a crash.
We're in a tight corner here. :)

File: src/buffer/out/OutputCellIterator.cpp [hunk 4]
```diff
@@ -97,14 +97,14 @@ OutputCellIterator::OutputCellIterator(const std::wstring_view utf16Text) :
 // Arguments:
 // - utf16Text - UTF-16 text range
 // - attribute - Color to apply over the entire range
-OutputCellIterator::OutputCellIterator(const std::wstring_view utf16Text, const TextAttribute attribute) :
+OutputCellIterator::OutputCellIterator(const std::wstring_view utf16Text, const TextAttribute& attribute, const size_t fillLimit) :
```
Comment (DHowett) [inline src/buffer/out/OutputCellIterator.cpp:100]: wat

Comment (lhecker) [reply src/buffer/out/OutputCellIterator.cpp:100]: I needed that so that I can write a unit test with surrogate characters.
The other tests only pass a single `wchar_t`...

**Review by DHowett — APPROVED** [review]

File: src/buffer/out/OutputCellIterator.cpp (hunk 4 above)
Comment (DHowett) [reply src/buffer/out/OutputCellIterator.cpp:100]: l o l
